yarp/yarp_compiler.c

```diff
@@ -97,19 +373,140 @@ yp_compile_node(rb_iseq_t *iseq, const yp_node_t *node, LINK_ANCHOR *const ret,
+          ADD_INSN1(ret, &dummy_line_node, newarraykwsplat, INT2FIX(0));
+
+          if (popped) {
+              ADD_INSN(ret, &dummy_line_node, pop);
+          }
+          return;
+      }
       case YP_NODE_BEGIN_NODE: {
           yp_begin_node_t *begin_node = (yp_begin_node_t *) node;
           if (begin_node->statements) {
               yp_compile_node(iseq, (yp_node_t *)begin_node->statements, ret, src, popped, compile_context);
           }
           return;
       }
+      case YP_NODE_CALL_NODE: {
+          yp_call_node_t *call_node = (yp_call_node_t *) node;
+
+          ID mid = parse_location_symbol(&call_node->message_loc);
+          int flags = 0;
+          int orig_argc = 0;
+
+          if (call_node->receiver == NULL) {
+              ADD_INSN(ret, &dummy_line_node, putself);
+          } else {
+              yp_compile_node(iseq, call_node->receiver, ret, src, popped, compile_context);
+          }
+
+          if (call_node->arguments == NULL) {
+              if (flags & VM_CALL_FCALL) {
+                  flags |= VM_CALL_VCALL;
+              }
+          } else {
+              yp_arguments_node_t *arguments = call_node->arguments;
+              yp_compile_node(iseq, (yp_node_t *) arguments, ret, src, popped, compile_context);
+              orig_argc = (int)arguments->arguments.size;
+          }
+
+          VALUE block_iseq = Qnil;
+          if (call_node->block != NULL) {
+              // Scope associated with the block
+              yp_scope_node_t scope_node;
+              yp_scope_node_init((yp_node_t *)call_node->block, &scope_node);
+
+              const rb_iseq_t *block_iseq = NEW_CHILD_ISEQ(&scope_node, make_name_for_block(iseq), ISEQ_TYPE_BLOCK, lineno);
+              ISEQ_COMPILE_DATA(iseq)->current_block = block_iseq;
+              ADD_SEND_WITH_BLOCK(ret, &dummy_line_node, mid, INT2FIX(orig_argc), block_iseq);
+          }
+          else {
+              if (block_iseq == Qnil && flags == 0) {
+                  flags |= VM_CALL_ARGS_SIMPLE;
+              }
+
+              if (call_node->receiver == NULL) {
+                  flags |= VM_CALL_FCALL;
+
+                  if (block_iseq == Qnil && call_node->arguments == NULL) {
+                      flags |= VM_CALL_VCALL;
+                  }
+              }
+
+              ADD_SEND_WITH_FLAG(ret, &dummy_line_node, mid, INT2NUM(orig_argc), INT2FIX(flags));
+          }
+          return;
+      }
+      case YP_NODE_CLASS_NODE: {
+          yp_class_node_t *class_node = (yp_class_node_t *)node;
+          yp_scope_node_t scope_node;
+          yp_scope_node_init((yp_node_t *)class_node, &scope_node);
+
+          yp_node_t *recursing_node = class_node->constant_path;
+
+          while (recursing_node->type != YP_NODE_CONSTANT_READ_NODE) {
+              recursing_node = ((yp_constant_path_node_t *)recursing_node)->child;
+          }
+
+          ID class_id = parse_location_symbol(&recursing_node->location);
```

Comment:
Should be able to use the name field on the class for this.

Reply:
Ah yeah nice, I had done this one before we added that!